routers/web/auth/auth.go

```diff
@@ -163,7 +163,7 @@ func SignIn(ctx *context.Context) {
 		context.SetCaptchaData(ctx)
 	}
 
-	ctx.HTML(http.StatusOK, tplSignIn)
+	ctx.HTML(http.StatusUnauthorized, tplSignIn)
```

Comment:
Is there any website use 401 for a sign-in page?

Reply:
I don't know. Do you have another idea how to handle this?

```diff
-       ctx.HTML(http.StatusUnauthorized, tplSignIn)
+       status := http.StatusOK
+       if ctx.Flash.ErrorMsg != "" {
+               status = http.StatusUnauthorized
+       }
+       ctx.HTML(status, tplSignIn)
```

Reply:
I haven't looked into the problem carefully.

Maybe something could be tried:

1. Use some test to show how SSPI works (mock the vendor package to make the test runnable in Gitea's CI)
2. Base on 1, Separate the SSPI response code logic from middleware.go . Or, separate the SSPI login logic into a new handler: `/user/login/sspi`, then it could do anything it needs.

(Just some initial thoughts)

Reply:
> Use some test to show how SSPI works (mock the vendor package to make the test runnable in Gitea's CI)

How do you mock a vendor package?

Thought about 2 too, but I didn't want to duplicate other logic.

The SSPI workflow is (don't have a setup to test it):
1. Client sends request
2. Server responds with `WWW-Authenticate Negotiate`
3. Client responds with `Authorization xxx`
4. Server responds with requested ressource (or again 2. if there are multiple "layers" of authorization needed)

The requests and responds should not be visible to the user, they don't have to see content. The original code just renders the login page if an error occurs which stops the automatic flow. Then they see the normal login page and can try again/login with another method.

If we agree we don't need the login page bonus, we could just do something like:
```go
func (s *SSPI) Verify(req *http.Request, w http.ResponseWriter, store DataStore, sess SessionStore) (*user_model.User, error) {
	...
	userInfo, outToken, err := sspiAuth.Authenticate(req, w)
	if err != nil {
		sspiAuth.AppendAuthenticateHeader(w, outToken)
		w.WriteHeader(http.StatusUnauthorized)
		w.Write([]byte("SSPI auth failed. Visit /user/login to login."))
		return nil, err
	}
```
But that still is bad design for such a method.

Reply:
> How do you mock a vendor package?

->  Make SSPI auth mockable #27036